### PR TITLE
Prep patches for sharing with CoreOS CI

### PR DESCRIPTION
**Pull request by jlebon**
See individual commit messages, but essentially the end goal here is preparations for sharing things with [CoreOS CI](https://github.com/jlebon/coreos-ci). Specifically, I want CoreOS CI to use the exact same Jenkins build configs, i.e. the inputs to the S2I builder. However, the pipeline and CoreOS CI will have different JCasC configurations (provided via configmaps) and Jenkins app templates (`jenkins.yaml`).

### CODE REVIEW

**Comment by jlebon** [comment]
Though there's a bunch of other unrelated minor fixes and enhancements in there along the way.

**Comment by jlebon** [comment]
OK, I've split out the preliminary ones that are easier to review in https://github.com/coreos/fedora-coreos-pipeline/pull/198.

**Comment by jlebon** [comment]
Note this is also prep for reworking how jobs are created and managed here too!

**Comment by jlebon** [comment]
Re-tested this successfully now in `fedora-coreos-devel`! :heavy_check_mark:

**Comment by jlebon** [comment]
Rebased!

File: deploy [hunk 1]
```diff
@@ -132,37 +132,49 @@ def get_username():
...
-    params = [q for p in params for q in ['--param', p]]
+    for k, v in params.items():
+        print(f"  {k}={v}")
     print()
 
-    resources = subprocess.check_output(
-        ['oc', 'process', '--filename', 'manifests/pipeline.yaml'] + params)
-    return json.loads(resources)
+    def gen_param_args(selected):
+        return [q for k, v in params.items() if k in selected for q in ['--param', f'{k}={v}']]
```
Comment (dustymabe) [inline deploy]: I find this hard to follow - maybe splitting into two operations would make things clearer

File: manifests/jenkins-s2i.yaml [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+      annotations:
+        coreos.com/deploy-default: "true"
+  - kind: BuildConfig
+    apiVersion: v1
+    metadata:
+      name: jenkins
+      annotations:
+        coreos.com/deploy-default: "true"
+        coreos.com/jenkins-jobs-url: ${JENKINS_JOBS_URL}
+        coreos.com/jenkins-jobs-ref: ${JENKINS_JOBS_REF}
```
Comment (dustymabe) [inline manifests/jenkins-s2i.yaml]: we added these two annotations as well?

File: deploy (hunk 1 above)
Comment (jlebon) [reply deploy]: Done!

File: manifests/jenkins-s2i.yaml (hunk 2 above)
Comment (jlebon) [reply manifests/jenkins-s2i.yaml]: Yeah sorry, I should've made that a separate commit. I'm not a fan myself of sneaking things into blocks that initially just looks like it was a plain move. :)

I added a comment there. Let me know if it's not clear!

File: deploy [hunk 3]
```diff
@@ -154,13 +154,19 @@ def process_template(args):
         print(f"  {k}={v}")
     print()
 
-    def gen_param_args():
-        return [q for k, v in params.items() for q in ['--param', f'{k}={v}']]
+    def gen_param_args(selected):
+        selected_params = {(k, v) for k, v in params.items() if k in selected}
+        return [q for k, v in selected_params for q in ['--param', f'{k}={v}']]
```
Comment (dustymabe) [inline deploy]: I think the part that was confusing to me here was the double for loop inside a list comprehension.. here's my alternative that I can read better:

```
def gen_param_args(selected):                                             
    selected_params = {(k, v) for k, v in params.items() if k in selected}
    args = []                                                             
    for k, v in selected_params:                                          
        args += ['--param', f'{k}={v}']                                   
    return args                                                           
                                                                          

```

I'll understand if you don't want to make this change.

Comment (jlebon) [reply deploy]: Not sure where this stands on the idiomatic scale (I've definitely seen it used elsewhere), but it's really useful. Here's how to read this: when you see something like `[w for y in z for x in y for w in x]`, you can simply unroll it in that same order:

```python
for y in z:
  for x in y:
    for w in x:
      mylist.append(w)
```

**Comment by dustymabe** [comment]
LGTM - merge away
